src/core/core.tooltip.js

```diff
@@ -771,31 +795,48 @@ var exports = Element.extend({
 		ctx.fillStyle = vm.bodyFontColor;
 		helpers.each(vm.beforeBody, fillLineOfText);
 
-		var drawColorBoxes = vm.displayColors;
-		xLinePadding = drawColorBoxes ? (bodyFontSize + 2) : 0;
+		xLinePadding = drawColorBoxes && bodyAlign !== 'right'
+			? bodyAlign === 'center' ? (bodyFontSize / 2 + 1) : (bodyFontSize + 2)
+			: 0;
 
 		// Draw body lines now
 		helpers.each(body, function(bodyItem, i) {
-			var textColor = vm.labelTextColors[i];
+			pt.x = textX;
+			textColor = vm.labelTextColors[i];
 			ctx.fillStyle = textColor;
 			helpers.each(bodyItem.before, fillLineOfText);
 
 			helpers.each(bodyItem.lines, function(line) {
 				// Draw Legend-like boxes if needed
 				if (drawColorBoxes) {
+					textWidth = ctx.measureText(line).width;
+					switch (bodyAlign) {
+					case 'center':
+						pt.x = textX - textWidth / 2 - xLinePadding;
+						break;
+					case 'right':
+						pt.x = textX - textWidth - bodyFontSize - 2;
+						break;
+					default:
+						pt.x = textX;
+						break;
+					}
+
 					// Fill a white rect so that colours merge nicely if the opacity is < 1
 					ctx.fillStyle = vm.legendColorBackground;
 					ctx.fillRect(pt.x, pt.y, bodyFontSize, bodyFontSize);
 
 					// Border
 					ctx.lineWidth = 1;
-					ctx.strokeStyle = vm.labelColors[i].borderColor;
+					ctx.strokeStyle = labelColors[i].borderColor;
 					ctx.strokeRect(pt.x, pt.y, bodyFontSize, bodyFontSize);
 
 					// Inner square
-					ctx.fillStyle = vm.labelColors[i].backgroundColor;
+					ctx.fillStyle = labelColors[i].backgroundColor;
 					ctx.fillRect(pt.x + 1, pt.y + 1, bodyFontSize - 2, bodyFontSize - 2);
 					ctx.fillStyle = textColor;
+
+					pt.x = textX;
```

Comment:
Instead of modifying `pt.x` and use `textX` to cache the initial value, what about using a new `x` variable?

```javascript
helpers.each(bodyItem.lines, function(line) {
    if (drawColorBoxes) {
        switch (bodyAlign) {
        case 'center':
           x = pt.x - textWidth / 2 - xLinePadding;
        // ...
        default:
            x = pt.x;
        }

        // ...
       
        ctx.fillRect(x, pt.y, bodyFontSize, bodyFontSize);

        // ...
    }
}
```